Storage: basic storage configuration layer

reinkrul opened the pull request:
Required for implementing backups and Redis support.

Backup configuration for BBolt is there for demonstration purposes (of how to configure databases), will be actually implemented later on.

Fixes https://github.com/nuts-foundation/nuts-node/issues/1169

TODO:
- [x] Test coverage
- [x] Documentation (write it when we have alternative backup strategies and/or Redis support).

reinkrul commented:
Ready for first round of comments.

woutslakhorst reviewed (commented):
My initial thoughts on config and DB setup was that any specific config for a particular DB is handled by stoabs. The concerns for how a DB functions lie with stoabs, then the Nuts node only have to select which DB is used. If a new type of DB is supported by stoabs this should be available with a lib update and shouldn't require any code changes.

reinkrul commented:
> My initial thoughts on config and DB setup was that any specific config for a particular DB is handled by stoabs. The concerns for how a DB functions lie with stoabs, then the Nuts node only have to select which DB is used. If a new type of DB is supported by stoabs this should be available with a lib update and shouldn't require any code changes.

Not 100% sure which part you mean. Config or also operational features, such as backupping?

Depends, my initial thought was to keep `go-stoabs` as simple as possible, so just an API to provide access to multiple databases. Mapping database options (e.g. connection strings, file names, etc) to configuration (be it a map or an actual Go struct) _could_ be a concern of `go-stoabs`, but for the sake of simplicity of the library I left it in `nuts-node`. Otherwise part of the Nuts Node config would be injected into `go-stoabs` which is certainly possible, but would complicate the Nuts Node documentation (then we should refer to the `go-stoabs` documentation on how to configure it, making our docs more complicated).

On the same note, scheduling a backup for BBolt is a Nuts Node concern as well: it's not something you can abstract in `go-stoabs`'s API, since whether/how to backup is highly dependent on the underlying database.

reinkrul commented:
@woutslakhorst  Tried to move database-specific config to `go-stoabs`, but I think the added complexity is not worth it (especially at this moment in time).

woutslakhorst commented:
Too much to handle in a single PR. Let's make this part about supporting the example config (last example in issue). I think there will be many more PRs to get things the way we want it to be.

woutslakhorst commented on storage/engine.go:
```diff
@@ -83,6 +94,40 @@ func (e engine) Shutdown() error {
+	for _, curr := range e.config.Databases {
+		if e.isDatabaseRegistered(curr.Type) {
+			// TODO: Will be supported in future
+			return fmt.Errorf("multiple databases configured of type '%s' (which is not supported)", curr.Type)
```
incorrect error?

woutslakhorst reviewed (changes requested):
And waiting for tests/successful build and a rebase

reinkrul commented:
Requires https://github.com/nuts-foundation/go-stoabs/pull/9 to be merged first (and then update library).

reinkrul commented:
@woutslakhorst I reworked the configuration. It's much simpler now, but more hardcoded. It reflects the configuration example as proposed by https://github.com/nuts-foundation/nuts-node/issues/1169 with one (important) difference: you can't configure a specific database (e.g. BBolt or Redis) multiple times. For the sake of simplicity I just assume data is not going to be in multiple Redis servers at this moment.

I don't think we need true "dynamic" support for any database supported by `go-stoabs` (in other words: when we update the library, the database is automagically supported in the Nuts Node): in the forseeable future we only support 2 databases, and having some config in the Nuts Node for each specific database makes configuration (for the end user) simpler imo (good auto-generated documentation, configuration in line with the rest of the node). Striving for that level of integration while being dynamic isn't worth the added complexity, I think.

And there will be database-specific features (e.g. backups) and metadata (storage class) that will still be database specific, implemented in the Nuts Node regardless of where they're configured.

woutslakhorst reviewed (commented):
1st a suggestion on the backup feature

woutslakhorst commented on storage/bbolt.go:
```diff
@@ -0,0 +1,182 @@
+	// Directory specifies the directory in which the BBolt backup should be written.
+	Directory string `koanf:"directory"`
+	// Interval specifies the time between backups.
+	Interval time.Duration `koanf:"interval"`
```
How is koanf suppose to coerce "1h" to a time.Duration?

reinkrul replied:
At first I thought that too and implemented a custom type, but appears it is supported out of the box...

woutslakhorst commented on storage/bbolt.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+func (b bboltDatabase) getClass() Class {
+	return VolatileStorageClass
```
If it has a backup running, then it's not volatile anymore?

reinkrul replied:
Probably? But if you configure a backup for BBolt, but configure a Redis server as well I'd say we'd give precedence to Redis for storing non-volatile data?

I don't feel like implementing it at this moment, because we don't support per-store database binding (e.g. DAG in Redis, network connections in BBolt) and backup is currently all or nothing (so having the separate classes won't change anything).

Change when we implemented Redis.

woutslakhorst commented on storage/bbolt.go:
```diff
@@ -0,0 +1,182 @@
+	// Write backup to "store.db.work"
+	// Rename existing backup ("store.db") to "store.db.previous"
+	// Rename "store.db.work" to "store.db"
+	// Remove work file
```
You just renamed it

woutslakhorst commented on storage/engine.go:
```diff
@@ -83,6 +89,14 @@ func (e engine) Shutdown() error {
 	e.datadir = config.Datadir
+
+	data, _ := json.MarshalIndent(e.config, "  ", "  ")
+	println(string(data))
```
println

reinkrul commented on storage/bbolt.go:
```diff
@@ -0,0 +1,182 @@
+	// Directory specifies the directory in which the BBolt backup should be written.
+	Directory string `koanf:"directory"`
+	// Interval specifies the time between backups.
+	Interval time.Duration `koanf:"interval"`
```
Made an issue for this: https://github.com/nuts-foundation/nuts-node/issues/1223

woutslakhorst reviewed (changes requested):
It looks like the storage engine now operates on a single BBolt DB instead of the backwards compatible files that used to back each store? This would require a migration for all current data!

reinkrul commented:
> It looks like the storage engine now operates on a single BBolt DB instead of the backwards compatible files that used to back each store? This would require a migration for all current data!

Nope, it should use the same files as before (`$datadir/$engine/$store.db`). See `bboltDatabase.createStore`.
